Feature/sketcher hints comprehensive coverage 22282

longrackslabs opened the pull request:
<!-- Add comprehensive context-aware hints for Sketcher constraint tools to address feedback about missing and incomplete hint coverage. -->

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
Fixes #22282 "Sketcher: Input hints don't cover all workflows"

## Before and After Images
<!-- If your proposed changes affect the FreeCAD GUI, add before and after screenshots -->

**Before:** Constraint tools showed generic or missing hints that didn't guide users through alternative workflows.

**After:** All constraint tools now provide context-aware hints that remember user selections and guide them through the correct workflow.

### Test Cases with Specific Hints

#### ConstrainRadiam
- **Test Case**: Create circle/arc → Select ConstrainRadiam → Shows "pick circle or arc"

#### ConstrainAngle
- **Test Case 1 (Line + Line)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick second line or point"
- **Test Case 2 (Point + Edge + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick first edge" (context-aware: knows you picked a point)
  - Step 2: "pick second edge"
- **Test Case 3 (Edge + Point + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick second line or point" (context-aware: knows you picked a line)
  - Step 2: "pick second edge" (context-aware: knows you're in edge+point+edge workflow)

#### ConstrainPerpendicular
- **Test Case 1 (Edge + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick second edge" (context-aware: knows you picked an edge)
- **Test Case 2 (Point + Edge + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick first edge" (context-aware: knows you picked a point)
  - Step 2: "pick second edge"

#### ConstrainTangent
- **Test Case 1 (Edge + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick second edge or point" (context-aware: knows you picked an edge)
- **Test Case 2 (Point + Edge + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick first edge" (context-aware: knows you picked a point)
  - Step 2: "pick second edge"
- **Test Case 3 (Edge + Point + Edge)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick second edge or point" (context-aware: knows you picked an edge)
  - Step 2: "pick second edge" (context-aware: knows you're in edge+point+edge workflow)

#### ConstrainSymmetric
- **Test Case 1 (Edge + Point)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick symmetry point" (context-aware: knows you picked an edge)
- **Test Case 2 (Point + Edge + Point)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick edge or second point" (context-aware: knows you picked a point)
  - Step 2: "pick symmetry point" (context-aware: knows you're in point+edge+point workflow)
- **Test Case 3 (Point + Point + Edge/Point)**: 
  - Step 0: "pick edge or first point"
  - Step 1: "pick edge or second point" (context-aware: knows you picked a point)
  - Step 2: "pick symmetry line or point" (context-aware: knows you're in point+point+edge/point workflow)

## Changes Made
- **ConstrainRadiam**: Added missing hints for auto radius/diameter tool
- **ConstrainAngle**: Added context-aware hints for all workflows (line+line, point+edge+edge, edge+point+edge)
- **ConstrainPerpendicular**: Added context-aware hints for all workflows (edge+edge, point+edge+edge)
- **ConstrainTangent**: Added context-aware hints for all workflows (edge+edge, point+edge+edge, edge+point+edge)
- **ConstrainSymmetric**: Added context-aware hints for all workflows (edge+point, point+edge+point, point+point+edge/point)

## Technical Details
- Hints now remember user selections to provide appropriate guidance for each workflow
- Uses special cases in `getToolHints()` method similar to existing `ConstrainPointOnObject`
- Covers all alternative workflows mentioned in the original feedback

longrackslabs commented on src/Mod/Sketcher/Gui/CommandConstraints.cpp:
```diff
@@ -1232,14 +1322,22 @@ class DrawSketchHandlerGenConstraint: public DrawSketchHandler
+
+            {.commandName = "Sketcher_ConstrainAngle",
+            .selectionStep = 2,
+            .hints = {{QObject::tr("%1 pick second edge"), {Gui::InputHint::UserInput::MouseLeft}}}},
```
Thanks for the feedback! I've refactored all the duplicated hint strings into class-level constants. This eliminates the duplication and makes the code much more maintainable.

longrackslabs commented on src/Mod/Sketcher/Gui/CommandConstraints.cpp:
```diff
@@ -1127,6 +1127,96 @@ class DrawSketchHandlerGenConstraint: public DrawSketchHandler
+                return {{QObject::tr("%1 pick second edge"), {Gui::InputHint::UserInput::MouseLeft}}};
+            } else if (isVertex(selSeq[1].GeoId, selSeq[1].PosId)) {
+                // Edge + Point + Edge workflow
+                return {{QObject::tr("%1 pick second edge"), {Gui::InputHint::UserInput::MouseLeft}}};
```
done

longrackslabs commented on src/Mod/Sketcher/Gui/CommandConstraints.cpp:
```diff
@@ -1127,6 +1127,96 @@ class DrawSketchHandlerGenConstraint: public DrawSketchHandler
+    // Special case for Sketcher_ConstrainPerpendicular to generate context-aware hints
+    if (commandName == "Sketcher_ConstrainPerpendicular") {
+        if (selectionStep == 0) {
+            return {{QObject::tr("%1 pick edge or first point"), {Gui::InputHint::UserInput::MouseLeft}}};
```
done

longrackslabs commented:
@FEA-eng thanks for the look and for the detailed feedback! I've updated the PR to address these issues:

Sketcher Dimension: Fixed the premature "place dimension" hint - now shows context-aware hints based on selection state instead of jumping straight to "place dimension".

ConstrainDistanceX/Y: Fixed the hint to show "pick second point" (not "pick second point or edge") when a point is selected first, since only points are accepted at that stage.

ConstrainPerpendicular: Added the edge+point+edge workflow with proper context-aware hints.

Let me know if you see anything else, these tools and hints can be nuanced.

FEA-eng commented:
> The transform tool modifiers for Offset tool and other transformation tools were actually covered in PR #21840. So the hints for those are already implemented.

Yes, but the task panel options (with keyboard shortcuts) are not included. For example:

<img width="1517" height="823" alt="2025-09-09 10_35_30-Greenshot" src="https://github.com/user-attachments/assets/ac441064-20e2-4392-92ba-74ebefac0acb" />

Same with other transform tools.

I can make a separate issue if you want to fix it later though.

longrackslabs commented:
Yes please make a separate issue

On Mon, Sep 8, 2025 at 22:40 FEA-eng ***@***.***> wrote:

> *FEA-eng* left a comment (FreeCAD/FreeCAD#23735)
> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3269550513>
>
> The transform tool modifiers for Offset tool and other transformation
> tools were actually covered in PR #21840
> <https://github.com/FreeCAD/FreeCAD/pull/21840>. So the hints for those
> are already implemented.
>
> Yes, but the task panel options (with keyboard shortcuts) are not
> included. For example:
> 2025-09-09.10_35_30-Greenshot.png (view on web)
> <https://github.com/user-attachments/assets/ac441064-20e2-4392-92ba-74ebefac0acb>
>
> Same with other transform tools.
>
> I can make a separate issue if you want to fix it later though.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3269550513>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KURSQBVZPK6X3RPCBD3R2G6VAVCNFSM6AAAAACF3UEZ5SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENRZGU2TANJRGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

FEA-eng commented:
> Yes please make a separate issue

Ok, done: #23815

FEA-eng commented:
I compiled this and it all looks good apart from those small remarks:
1. Sketcher Dimension:
- previously it said "place dimension" too early, now it doesn't say it at all. Perhaps it would be good to just add "or place dimension" to "pick ..." when placing dimensions is possible instead of picking another geometrical entity.
- also, it should mention the M key as noted [here](https://github.com/FreeCAD/FreeCAD/issues/22282#issuecomment-3027268721).

2. Sketcher ConstrainSymmetric:
- the two points and symmetry point workflow is not covered, it says that symmetry line has to be selected after picking two points (so it should say "pick symmetry line or symmetry point")
- the point, symmetry line and another point workflow is not covered, it only says to pick edge or second point after picking first point and then when you pick and edge (symmetry line) it changes to "pick symmetry point" while it should mention regular point too

longrackslabs commented:
Fixed #2: Two points + symmetry point workflow
Updated hint to "pick symmetry line or symmetry point"
Added missing sequence {SelVertexOrRoot, SelVertexOrRoot, SelVertexOrRoot}
Implemented case 8 in applyConstraint() to handle three-point symmetry

Question about fall-through behavior: The code after the switch statement
creates a constraint for cases that don't explicitly return. This appears
to be intentional - it handles the original constraint creation logic for
cases that fall through (not sure what those might be). I am careful to
focus on my code so I did not do any additional changes or refactors of
that bit.


On Sat, Sep 20, 2025 at 2:36 AM FEA-eng ***@***.***> wrote:

> *FEA-eng* left a comment (FreeCAD/FreeCAD#23735)
> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3314857435>
>
> I compiled this and it all looks good apart from those small remarks:
>
>    1. Sketcher Dimension:
>
>
>    - previously it said "place dimension" too early, now it doesn't say
>    it at all. Perhaps it would be good to just add "or place dimension" to
>    "pick ..." when placing dimensions is possible instead of picking another
>    geometrical entity.
>    - also, it should mention the M key as noted here
>    <https://github.com/FreeCAD/FreeCAD/issues/22282#issuecomment-3027268721>
>    .
>
>
>    2. Sketcher ConstrainSymmetric:
>
>
>    - the two points and symmetry point workflow is not covered, it says
>    that symmetry line has to be selected after picking two points (so it
>    should say "pick symmetry line or symmetry point")
>    - the point, symmetry line and another point workflow is not covered,
>    it only says to pick edge or second point after picking first point and
>    then when you pick and edge (symmetry line) it changes to "pick symmetry
>    point" while it should mention regular point too
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3314857435>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KX6M2DJ6LLUNRWZA6D3TUNYNAVCNFSM6AAAAACF3UEZ5SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMJUHA2TONBTGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

longrackslabs commented:
On comment #1, I've added the M key mode hint to the dimension tool. The
dimension tool now shows "switch mode" when the M key is available for mode
switching. I focused on adding the hint itself, not modifying how the mode
switching actually works.

I am sorry for being dense, but I don't understand this other comment

   - previously it said "place dimension" too early, now it doesn't say it
   at all. Perhaps it would be good to just add "or place dimension" to "pick
   ..." when placing dimensions is possible instead of picking another
   geometrical entity."



On Sat, Sep 20, 2025 at 11:29 AM Longracks Labs ***@***.***>
wrote:

> Fixed #2: Two points + symmetry point workflow
> Updated hint to "pick symmetry line or symmetry point"
> Added missing sequence {SelVertexOrRoot, SelVertexOrRoot, SelVertexOrRoot}
> Implemented case 8 in applyConstraint() to handle three-point symmetry
>
> Question about fall-through behavior: The code after the switch statement
> creates a constraint for cases that don't explicitly return. This appears
> to be intentional - it handles the original constraint creation logic for
> cases that fall through (not sure what those might be). I am careful to
> focus on my code so I did not do any additional changes or refactors of
> that bit.
>
>
> On Sat, Sep 20, 2025 at 2:36 AM FEA-eng ***@***.***> wrote:
>
>> *FEA-eng* left a comment (FreeCAD/FreeCAD#23735)
>> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3314857435>
>>
>> I compiled this and it all looks good apart from those small remarks:
>>
>>    1. Sketcher Dimension:
>>
>>
>>    - previously it said "place dimension" too early, now it doesn't say
>>    it at all. Perhaps it would be good to just add "or place dimension" to
>>    "pick ..." when placing dimensions is possible instead of picking another
>>    geometrical entity.
>>    - also, it should mention the M key as noted here
>>    <https://github.com/FreeCAD/FreeCAD/issues/22282#issuecomment-3027268721>
>>    .
>>
>>
>>    2. Sketcher ConstrainSymmetric:
>>
>>
>>    - the two points and symmetry point workflow is not covered, it says
>>    that symmetry line has to be selected after picking two points (so it
>>    should say "pick symmetry line or symmetry point")
>>    - the point, symmetry line and another point workflow is not covered,
>>    it only says to pick edge or second point after picking first point and
>>    then when you pick and edge (symmetry line) it changes to "pick symmetry
>>    point" while it should mention regular point too
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3314857435>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAI67KX6M2DJ6LLUNRWZA6D3TUNYNAVCNFSM6AAAAACF3UEZ5SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMJUHA2TONBTGU>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

maxwxyz commented:
The PR has a conflict

longrackslabs commented:
Conflicts resolved and all checks passing. Ready for review.

FEA-eng commented:
Thanks for the changes. IMO only this is left:
- "place dimension" for Sketcher Dimension - when using this tool, it's always possible to just click on the background instead of picking another geometrical entity and this places the dimension label attached to the cursor in the selected spot. So tooltips could always include "or place dimension". But we don't have to go this way if we decide that it would make the hints for this tool too cluttered.
- the point, symmetry line and another point workflow for Constraint Symmetric - now when I pick the first point, it says "pick edge or second point". Even if we treat "edge" as "symmetry line", after picking it, the hints says "pick symmetry point" while it's not correct for this workflow - the last point is a regular point.

longrackslabs commented:
Good point about the dimension tool. I was thinking about the 'place
dimension' wording - since users aren't really placing it somewhere
specific ( thats what i thought you meant by 'place') - but rather
finishing the dimension to set its value, would 'or click to finish' be
clearer? Also, this option only becomes available after picking the first
geometry. Let me know if you'd like me to implement this or if we should
skip it to avoid clutter.

On Sun, Sep 21, 2025 at 1:09 PM FEA-eng ***@***.***> wrote:

> *FEA-eng* left a comment (FreeCAD/FreeCAD#23735)
> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3316235284>
>
> Thanks for the changes. IMO only this is left:
>
>    - "place dimension" for Sketcher Dimension - when using this tool,
>    it's always possible to just click on the background instead of picking
>    another geometrical entity and this places the dimension label attached to
>    the cursor in the selected spot. So tooltips could always include "or place
>    dimension". But we don't have to go this way if we decide that it would
>    make the hints for this tool too cluttered.
>    - the point, symmetry line and another point workflow for Constraint
>    Symmetric - now when I pick the first point, it says "pick edge or second
>    point". Even if we treat "edge" as "symmetry line", after picking it, the
>    hints says "pick symmetry point" while it's not correct for this workflow -
>    the last point is a regular point.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/23735#issuecomment-3316235284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KQ6TIBITWFNQTDCSBL3T4AZHAVCNFSM6AAAAACF3UEZ5SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMJWGIZTKMRYGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

FEA-eng commented:
> would 'or click to finish' be clearer?

Yes, I would go with this.

maxwxyz commented on src/Mod/Sketcher/Gui/CommandConstraints.cpp:
```diff
@@ -1575,6 +1733,13 @@ class GeomSelectionSizes
+    static constexpr const char* PICK_EDGE = "%1 pick edge";
+    static constexpr const char* PICK_POINT_OR_EDGE = "%1 pick point or edge";
+    static constexpr const char* PICK_SECOND_POINT_OR_EDGE = "%1 pick second point or edge";
+    static constexpr const char* PICK_SECOND_POINT_OR_EDGE_OR_CLICK_TO_FINISH = "%1 pick second point or edge, or click to finish";
```
", or click to finish" is a bad hint name, as we also do not display e.g. M - click M to toggle ("click to" is not needed as it is implied by the hint itself.
Finish in this case is cancel, ok or place the dimension? IMO this is not clear.

longrackslabs commented:
I’m going to step back from this one — I’ve taken it as far as I can, and I don’t want to keep circling. 

Thanks for the feedback and discussion. If anyone wants to pick this up and push it further, please feel free to build on this work.

FEA-eng reviewed (approved):
For me, this is sufficient. IMO the PR should be merged as it fixes the referenced issue (adds missing hints). Wording improvements can be done in another PR in the future once DWG decides on the nomenclature. Especially since the existing hints would also need such changes.

longrackslabs commented:
Thank you.

chennes commented:
@maxwxyz please let me know when this is ready to go.

chennes commented:
OK it looks like something went sideways during a merge or a rebase. I think I've gotten it all cleaned up, but before I force-push the cleanup to your branch could you please verify that this branch has all of the commits you expect to see, @longrackslabs ?
https://github.com/FreeCAD/FreeCAD/compare/main...chennes:FreeCAD:cleanup_23735
